File: Live Class - Python/rate_limiter.py
#Slide 50 : Rate Limit Implementation
import time
import threading
from functools import wraps

class RateLimiter:
  def __init__(self, max_calls, period):
    self.max_calls = max_calls
    self.period = period
    self.counter = 0
    self.lock = threading.Lock()
    self.last_reset_time = time.time()

  def is_allowed(self):
    current_time = time.time()

    with self.lock:
      # Check if the time period has passed since the last reset
      if current_time - self.last_reset_time > self.period:
        self.counter = 0
        self.last_reset_time = current_time

      if self.counter < self.max_calls:
        self.counter += 1
        return True
      else:
        return False

  def limit(self, func):
    @wraps(func)
    def wrapper(*args, **kwargs):
      if self.is_allowed():
        return func(*args, **kwargs)
      else:
        print(f"Rate limit exceeded. Try again later.")
    return wrapper
    
rate_limiter = RateLimiter(max_calls=5, period=1)

@rate_limiter.limit
def my_function():
  print(f"Function executed by {threading.current_thread().name}")

if __name__ == "__main__":
  def worker():
    for i in range(10):
      my_function()
      time.sleep(0.2)  # Simulate work being done in the function

  # Create multiple threads to simulate concurrent function calls
  threads = [threading.Thread(target=worker, name=f"Thread-{i}") for i in range(3)]

  # Start all threads
  for thread in threads:
    thread.start()

  # Wait for all threads to finish
  for thread in threads:
    thread.join()

File: Live Class - Java/Rate_Limiter.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class RateLimiter {
    private final int maxCalls;
    private final long period;
    private int counter;
    private long lastResetTime;
    private final Lock lock = new ReentrantLock();

    public RateLimiter(int maxCalls, long period) {
        this.maxCalls = maxCalls;
        this.period = period * 1000; // Convert seconds to milliseconds
        this.counter = 0;
        this.lastResetTime = System.currentTimeMillis();
    }

    public boolean isAllowed() {
        long currentTime = System.currentTimeMillis();

        lock.lock();
        try {
            // Check if the time period has passed since the last reset
            if (currentTime - lastResetTime > period) {
                counter = 0;
                lastResetTime = currentTime;
            }

            if (counter < maxCalls) {
                counter++;
                return true;
            } else {
                return false;
            }
        } finally {
            lock.unlock();
        }
    }

    public void limit(Runnable task) {
        if (isAllowed()) {
            task.run();
        } else {
            System.out.println("Rate limit exceeded. Try again later.");
        }
    }
}

public class Rate_Limiter {
    private static final RateLimiter rateLimiter = new RateLimiter(5, 1);

    public static void myFunction() {
        System.out.println("Function executed by " + Thread.currentThread().getName());
    }

    public static void worker() {
        for (int i = 0; i < 10; i++) {
            rateLimiter.limit(Rate_Limiter::myFunction);
            try {
                Thread.sleep(200); // Simulate work being done in the function
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    public static void main(String[] args) {
        // Create multiple threads to simulate concurrent function calls
        Thread[] threads = new Thread[3];
        for (int i = 0; i < threads.length; i++) {
            threads[i] = new Thread(Rate_Limiter::worker, "Thread-" + i);
        }

        // Start all threads
        for (Thread thread : threads) {
            thread.start();
        }

        // Wait for all threads to finish
        for (Thread thread : threads) {
            try {
                thread.join();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
